File: games/shell_template/native-resources/ios/App/AppsFlyerAttributionPlugin.swift
import AppsFlyerLib
import Capacitor
import UIKit

@objc(AppsFlyerAttributionPlugin)
public class AppsFlyerAttributionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppsFlyerAttributionPlugin"
    public let jsName = "AppsFlyerAttribution"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
    ]

    private var initialized = false

    @objc func initialize(_ call: CAPPluginCall) {
        onMain { [weak self] in
            self?.initializeOnMain(call)
        }
    }

    @objc func trackEvent(_ call: CAPPluginCall) {
        onMain { [weak self] in
            self?.trackEventOnMain(call)
        }
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        onMain { [weak self] in
            guard let self else {
                call.resolve(["initialized": false, "appsFlyerId": NSNull()])
                return
            }
            var status: JSObject = ["initialized": self.initialized]
            if self.initialized {
                status["appsFlyerId"] = AppsFlyerLib.shared().getAppsFlyerUID()
            } else {
                status["appsFlyerId"] = NSNull()
            }
            call.resolve(status)
        }
    }

    private func initializeOnMain(_ call: CAPPluginCall) {
        guard let devKey = trimmedString(call.getString("devKey")) else {
            log("initialize skipped; missing dev key")
            call.resolve(["initialized": false])
            return
        }
        guard let appleAppId = trimmedString(call.getString("appleAppId")) else {
            log("initialize skipped; missing apple app id")
            call.resolve(["initialized": false])
            return
        }
        if initialized {
            call.resolve(["initialized": true])
            return
        }

        let lib = AppsFlyerLib.shared()
        lib.appsFlyerDevKey = devKey
        lib.appleAppID = appleAppId
        lib.isDebug = call.getBool("debugLogging") ?? false
        let attWaitSeconds = call.getInt("attWaitSeconds") ?? 0
        if attWaitSeconds > 0 {
            // Delay the first launch report until ATT resolves (or the timeout
            // fires) so conversion data carries the user's real consent state.
            lib.waitForATTUserAuthorization(timeoutInterval: Double(attWaitSeconds))
        }
        lib.start()
        initialized = true
        log("AppsFlyer SDK started appleAppId=\(appleAppId) devKey=\(redacted(devKey))")
        call.resolve(["initialized": true])
    }

    private func trackEventOnMain(_ call: CAPPluginCall) {
        guard initialized else {
            log("event skipped; SDK not initialized")
            call.resolve(["tracked": false])
            return
        }
        guard let eventName = trimmedString(call.getString("eventName")) else {
            log("event skipped; missing event name")
            call.resolve(["tracked": false])
            return
        }

        var values: [String: Any] = [:]
        for (key, value) in call.getObject("eventValues") ?? [:] {
            guard let stringValue = value as? String else { continue }
            values[key] = capped(stringValue)
        }
        AppsFlyerLib.shared().logEvent(eventName, withValues: values)
        log("event logged name=\(eventName)")
        call.resolve(["tracked": true])
    }

    private func trimmedString(_ value: String?) -> String? {
        guard let value else {
            return nil
        }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private func capped(_ value: String) -> String {
        if value.count <= 96 {
            return value
        }
        return String(value.prefix(96))
    }

    private func onMain(_ work: @escaping () -> Void) {
        if Thread.isMainThread {
            work()
        } else {
            DispatchQueue.main.async(execute: work)
        }
    }

    private func redacted(_ value: String) -> String {
        if value.count <= 6 {
            return "<redacted>"
        }
        return "<redacted:\(value.suffix(4))>"
    }

    private func log(_ message: String) {
        NSLog("[attribution:appsflyer] \(message)")
    }
}

File: games/find_the_dog/native-resources/android/app/src/main/java/com/basegamelab/findthedog/sdk/AppsFlyerAttributionPlugin.java
package com.basegamelab.findthedog.sdk;

import com.appsflyer.AppsFlyerLib;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import org.json.JSONObject;

@CapacitorPlugin(name = "AppsFlyerAttribution")
public class AppsFlyerAttributionPlugin extends Plugin {
    private boolean initialized = false;

    @PluginMethod
    public void initialize(PluginCall call) {
        String devKey = trimmed(call.getString("devKey"));
        if (devKey == null) { resolve(call, "initialized", false); return; }
        if (!initialized) {
            AppsFlyerLib lib = AppsFlyerLib.getInstance();
            lib.setDebugLog(Boolean.TRUE.equals(call.getBoolean("debugLogging", false)));
            lib.init(devKey, null, getContext().getApplicationContext());
            lib.start(getContext().getApplicationContext());
            initialized = true;
        }
        resolve(call, "initialized", true);
    }

    @PluginMethod
    public void trackEvent(PluginCall call) {
        String eventName = trimmed(call.getString("eventName"));
        if (!initialized || eventName == null) { resolve(call, "tracked", false); return; }
        Map<String, Object> values = new HashMap<>();
        JSONObject raw = call.getObject("eventValues", new JSObject());
        Iterator<String> keys = raw.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            Object value = raw.opt(key);
            if (value instanceof String) values.put(key, capped((String) value));
        }
        AppsFlyerLib.getInstance().logEvent(getContext().getApplicationContext(), eventName, values);
        resolve(call, "tracked", true);
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        JSObject status = new JSObject();
        status.put("initialized", initialized);
        status.put("appsFlyerId", initialized ? AppsFlyerLib.getInstance().getAppsFlyerUID(getContext()) : JSObject.NULL);
        call.resolve(status);
    }

    private static void resolve(PluginCall call, String key, boolean value) { JSObject result = new JSObject(); result.put(key, value); call.resolve(result); }
    private static String trimmed(String value) { if (value == null) return null; String out = value.trim(); return out.isEmpty() ? null : out; }
    private static String capped(String value) { return value.length() <= 96 ? value : value.substring(0, 96); }
}
